src/client/gpm_common.c

```diff
@@ -300,55 +299,41 @@ static int gpm_epoll_wait(struct gpm_ctx *gpmctx, uint32_t event_flags)
     if (epoll_ret < 0) {
         /* Error while waiting that isn't EINTR */
         ret = errno;
-        gpm_epoll_close(gpmctx);
+        goto done;
     } else if (epoll_ret == 0) {
         ret = ETIMEDOUT;
-        gpm_epoll_close(gpmctx);
-    } else if (epoll_ret == 1 && events[0].data.fd == gpmctx->timerfd) {
-        /* Got an event which is only our timer */
-        if ((events[0].events & EPOLLIN) == 0) {
-            /* We got an event which was not EPOLLIN; assume this is an error,
-             * and exit with EBADF: epoll_wait said timerfd had an event,
-             * but that event is not an EPOLIN event. */
-            ret = EBADF;
-        } else {
-            ret = read(gpmctx->timerfd, &timer_read, sizeof(uint64_t));
-            if (ret == -1 && errno != EAGAIN && errno != EWOULDBLOCK) {
-                /* In the case when reading from the timer failed, don't hide the
-                 * timer error behind ETIMEDOUT such that it isn't retried */
-                ret = errno;
-            } else {
-                /* If ret == 0, then we definitely timed out. Else, if ret == -1
-                 * and errno == EAGAIN or errno == EWOULDBLOCK, we're in a weird
-                 * edge case where epoll thinks the timer can be read, but it
-                 * is blocking more; treat it like a TIMEOUT and retry, as
-                 * nothing around us would handle EAGAIN from timer and retry
-                 * it. */
-                ret = ETIMEDOUT;
-            }
-        }
-        gpm_epoll_close(gpmctx);
-    } else {
-        /* If ret == 2, then we ignore the timerfd; that way if the next
-         * operation cannot be performed immediately, we timeout and retry.
-         * Always check the returned event of the socket fd. */
-        int fd_index = 0;
-        if (epoll_ret == 2 && events[fd_index].data.fd != gpmctx->fd) {
-            fd_index = 1;
+        goto done;
+    } else if (epoll_ret == 2 || events[0].data.fd == gpmctx->timerfd ||
+               events[1].data.fd == gpmctx->timerfd) {
+        /* Timer fired - service it. */
+        ret = read(gpmctx->timerfd, &timer_read, sizeof(uint64_t));
+        if (ret == -1 && errno != EAGAIN && errno != EWOULDBLOCK) {
+            /* In the case when reading from the timer failed, don't hide the
+             * timer error behind ETIMEDOUT such that it isn't retried */
+            ret = errno;
+            goto done;
         }
 
-        if ((events[fd_index].events & event_flags) == 0) {
-            /* We cannot call EPOLLIN/EPOLLOUT at this time; assume that this
-             * is a fatal error; return with EBADFD to distinguish from
-             * EBADF in timer_fd case. */
-            ret = EBADFD;
-            gpm_epoll_close(gpmctx);
-        } else {
-            /* We definintely got a EPOLLIN/EPOLLOUT event; return success. */
-            ret = 0;
-        }
+        /* If ret == 0, then we definitely timed out. Else, if ret == -1 and
```

Comment:
Aha! I think this is another subtle bug. Technically, under some interpretation of the system's `read` man page, because `read` _can_ return `ret < sizeof(uint64_t)`, (say, if we read some bytes and then hit an interrupt), rather than returning `EINTR`, we can return `ret > 0 && ret < sizeof(uint64_t)`, and technically we'd need to retry the `read` until we hit the 8 bytes that the timer returns.

Reply:
This might be a weird edge case on some systems? And might perhaps cause a loop, because the timer will still be ready? But eventually we'd read those 8 bytes, hmm.... Unless it eventually exits `gpm_read_...` and we close the `timerfd` and `epollfd` and re-create the entire thing? Hmmm... 

Also, for a weird edge case, could we end up with _two_ events? :-)

Reply:
Aha!! Is this where the bug is?

See, when we read `timerfd` but could also read `gpmctx->fd`, I don't think we really want to return `ETIMEDOUT` here. We want to fall through, don't we? And let the `read`/`write` of `gpmctx->fd` succeed in the event we have `epoll_ret == 2`? And "just" reset the timer?

Reply:
Again, we never reset the timer.  Any data on the timerfd hangs up the connection.